lib/protoparser/prometheus/parser.go

```diff
@@ -727,3 +729,119 @@ var numericChars = [256]bool{
 	'E': true,
 	'.': true,
 }
+
+// Metadata contains HELP or TYPE message in prometheus exposition format.
+// See https://github.com/prometheus/docs/blob/e39897e4ee6e67d49d47204a34d120e3314e82f9/docs/instrumenting/exposition_formats.md#comments-help-text-and-type-information.
+//
+// For example:
+// # HELP alertmanager_alerts How many alerts by state.
+// # TYPE alertmanager_alerts gauge
+type Metadata struct {
+	Metric string
+	Type   uint32
+	Help   string
+}
+
+// MetadataRows contains parsed Prometheus metadata rows.
+type MetadataRows struct {
+	Rows []Metadata
+}
+
+// Unmarshal unmarshals Prometheus metadata rows from s with stdErrLogger.
+func (ms *MetadataRows) Unmarshal(s string) {
+	ms.UnmarshalWithErrLogger(s, stdErrLogger)
+}
+
+// UnmarshalWithErrLogger unmarshals Prometheus metadata rows from s.
+func (ms *MetadataRows) UnmarshalWithErrLogger(s string, errLogger func(s string)) {
+	ms.Rows = unmarshalMetadataRowsWithErrLogger(s, errLogger)
+}
+
+// Reset resets ms.
+func (ms *MetadataRows) Reset() {
+	ms.Rows = ms.Rows[:0]
+}
+
+func unmarshalMetadataRowsWithErrLogger(s string, errLogger func(s string)) []Metadata {
+	var dst []Metadata
+	for len(s) > 0 {
+		var line string
+		n := strings.IndexByte(s, '\n')
+		if n < 0 {
+			line = s
+			s = ""
+		} else {
+			line = s[:n]
+			s = s[n+1:]
+		}
+		if mr, ok := parseMetadataLine(line, errLogger); ok {
+			dst = append(dst, mr)
+		}
+	}
+	metadataReadScrape.Add(len(dst))
+	return dst
+}
+
+func parseMetadataLine(s string, errLogger func(s string)) (Metadata, bool) {
+	s = skipLeadingWhitespace(s)
+
+	if len(s) < 2 || s[0] != '#' || s[1] != ' ' {
+		// Skip non-comment
+		return Metadata{}, false
+	}
+
+	parts := strings.SplitN(s, " ", 4)
+	if len(parts) < 3 {
+		if errLogger != nil {
+			errLogger(fmt.Sprintf("cannot unmarshal metadata line %q", s))
+		}
+		return Metadata{}, false
+	}
+	var mr Metadata
+	commentType := parts[1]
+	mr.Metric = strings.Clone(parts[2])
+	var isType, isHelp bool
+	switch commentType {
+	case "HELP":
+		isHelp = true
+	case "TYPE":
+		isType = true
+	default:
+		if errLogger != nil {
+			errLogger(fmt.Sprintf("cannot unmarshal metadata line %q", s))
```

Comment:
Prometheus silently ignores metadata parsing errors.

Reply:
The current scrape error handling between Prometheus and VictoriaMetrics are different:
1. Prometheus aborts the scrape on any parse error (metadata or samples), marks it as failed with the error message, and logs nothing, see https://github.com/prometheus/prometheus/blob/2c04f2d7b1a7bbd260ffae70d1bd57b20c0c0d4f/scrape/scrape.go#L1482-L1487
2. VictoriaMetrics only logs the parse error, skips the problematic line and considers the scrape successful if the request succeeded, see
https://github.com/VictoriaMetrics/VictoriaMetrics/blob/65251c8fe27250b8ef031c550747ebecd28a245a/lib/protoparser/prometheus/parser.go#L224

To be consistent, metadata parsing should also follow the current pattern, log the error and continue.